test-case-reporting/src/test_result_record.ts

```diff
@@ -120,89 +123,109 @@ export class TestResultRecord {
-   * @param buildNumber The number of the Travis build, as a string
-   */
-  private async getBuildFromBuildNumber(buildNumber: string): Promise<Build> {
-    const dbBuild = await this.getDbBuildFromBuildNumber(buildNumber);
-    return this.getBuildFromDbBuild(dbBuild);
+    const testCase: TestCase = {
+      name,
+      createdAt: getDateFromTimestamp(created_at),
+    };
+
+    const testRun: TestRun = {
+      job,
+      testCase,
+      status,
+      timestamp: getDateFromTimestamp(timestamp),
+      durationMs: duration_ms,
+    };
+
+    return testRun;
   }
 
-  private async getTestRunsOfBuild(
-    buildNumber: string
+  private async bigJoinQuery(
+    queryFunction: QueryFunction,
+    {limit, offset}: PageInfo
   ): Promise<Array<TestRun>> {
-    const build: Build = await this.getBuildFromBuildNumber(buildNumber);
-
-    const dbJoins: Array<DB.BigJoin> = await this.db<DB.BigJoin>('builds')
-      .where('build_number', buildNumber)
+    const baseQuery = this.db<DB.BigJoin>('builds')
       .join('jobs', 'jobs.build_id', 'builds.id')
       .join('test_runs', 'tests_runs.job_id', 'jobs.id')
       .join('test_cases', 'test_cases.id', 'test_runs.test_case_id')
-      .select(
-        'jobs.job_number',
-        'jobs.test_suite_type',
-
-        'test_cases.name',
-        'test_cases.created_at',
-
-        'test_runs.status',
-        'test_runs.timestamp',
-        'test_runs.duration_ms'
-      );
-
-    const testRuns: Array<TestRun> = dbJoins.map(
-      ({
-        job_number,
-        test_suite_type,
-        name,
-        created_at,
-        status,
-        timestamp,
-        duration_ms,
-      }: DB.BigJoin) => ({
-        job: {
-          build,
-          jobNumber: job_number,
-          testSuiteType: test_suite_type,
-        },
-
-        testCase: {
-          name,
-          createdAt: getDateFromTimestamp(created_at),
-        },
-
-        status,
-        timestamp: getDateFromTimestamp(timestamp),
-        durationMs: duration_ms,
-      })
+      .limit(limit)
+      .offset(offset);
+
+    const fullQuery = queryFunction(baseQuery);
+
+    const rows = await fullQuery.select(
+      'builds.build_number',
+      'builds.commit_sha',
+
+      'jobs.job_number',
+      'jobs.test_suite_type',
+
+      'test_cases.name',
+      'test_cases.created_at',
+
+      'test_runs.status',
+      'test_runs.timestamp',
+      'test_runs.duration_ms'
     );
 
-    return testRuns;
+    return rows.map(this.getTestRunFromRow);
+  }
+
+  /**
+   * Gets a list of the test results belonging to a build
+   * @param buildNumber The number of the Travis build whose test runs we want.
+   */
+  async getTestRunsOfBuild(
+    buildNumber: string,
+    {limit, offset}: PageInfo
+  ): Promise<Array<TestRun>> {
+    const queryFunction: QueryFunction = (q: QueryBuilder.QueryBuilder) =>
+      q.where('build_number', buildNumber);
+
+    return this.bigJoinQuery(queryFunction, {
+      limit,
+      offset,
+    });
```

Comment:
Question: Does Typescript complain if this is shortened, like

```javascript
return this.bigJoinQuery(
  q => q.where('build_number', buildNumber),
  pageInfo
)
```
I'm guessing yes, which makes me :disappointed:

Reply:
It does :(
Typescript gets real verbose with callbacks & the like...

Reply:
Scratch that, I was importing the function type incorrectly. We can omit the explicit type for the argument :)

Done.